File: common/src/main/java/com/yekong/common/baseadapter/BaseAdapter.java
package com.yekong.common.baseadapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

/**
 * M为数据类型
 * Created by xigua on 2017/10/17.
 */

public abstract class BaseAdapter<M> extends RecyclerView.Adapter {
    public ItemClickListener<M> listener;
    public Context context;
    private List<M> list;

    public BaseAdapter(ItemClickListener listener, Context context , List<M> list) {
        this.listener = listener;
        this.context = context;
        this.list = list;
    }

    public void setList(List<M> list) {
        this.list = list;
        notifyDataSetChanged();
    }

    public List<M> getList(){return list;}

    @Override
    public int getItemViewType(int position) {
        return getViewType(position);
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return getViewHolder(context,parent,viewType);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        onBindView(holder,position,list.get(position));
    }

    public View getView(int id){
        return View.inflate(context, id, null);
    }

    @Override
    public int getItemCount() {
        return list == null?0:list.size();
    }
    public abstract RecyclerView.ViewHolder getViewHolder(Context context , ViewGroup parent , int viewType);
    public abstract void onBindView(RecyclerView.ViewHolder h , int p , M m);
    public abstract int getViewType(int position);
}

File: kotlin_common/src/main/java/com/yekong/kotlin_common/baseadapter/BaseAdapter.kt
package com.yekong.kotlin_common.baseadapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup

/**
 * Created by xigua on 2017/10/19.
 */
abstract class BaseAdapter <M>(var context: Context, var listener: ItemClickListener<M>, var list: List<M>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
        onBind(holder,position,list.get(position))
    }

    override fun getItemCount(): Int {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
        return if (list == null) 0 else list.size
    }

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): RecyclerView.ViewHolder {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
       return getHolder()
    }

    override fun getItemViewType(position: Int): Int {
        return getType()
    }

    fun setData(list: List<M>){
        this.list = list
        notifyDataSetChanged()
    }

    abstract fun getHolder():RecyclerView.ViewHolder
    abstract fun onBind(holder: RecyclerView.ViewHolder, position: Int,m : M)
    abstract fun getType():Int

}